packages/java-edition/src/json/checker/index.ts

```diff
@@ -82,269 +80,59 @@ export function definition(
+					case 'short':
+					case 'int':
+					case 'long':
+						return [ 'byte', 'short', 'int', 'long', 'float', 'double' ].includes(def.kind);
+					case 'float':
+					case 'double':
+						return [ 'float', 'double' ].includes(def.kind);
+					default: return false;
 				}
-				if (type.valueRange) {
-					for (const { value: childNode } of node.children) {
-						if (childNode?.type !== 'json:number') {
-							ctx.err.report(
-								localize('expected', localizeTag('json:number')),
-								node,
-								core.ErrorSeverity.Warning,
-							)
-						} else if (
-							childNode &&
-							!isInRange(Number(childNode.value), type.valueRange)
-						) {
-							ctx.err.report(
-								localize(
-									'number.between',
-									type.valueRange.min ?? '-∞',
-									type.valueRange.max ?? '+∞',
-								),
-								node,
-								core.ErrorSeverity.Warning,
-							)
-						}
-					}
+			},
+			getChildren: node => {
+				if (node.type === 'json:array') {
+					return node.children.filter(n => n.value)
+						.map(n => ([{ originalNode: n.value!, inferredType: inferType(n.value!) }]))
 				}
-				break
-			case 'byte':
-			case 'short':
-			case 'int':
-			case 'long':
-			case 'float':
-			case 'double':
-				node = node as JsonNumberNode
-				if (
-					type.valueRange &&
-					!isInRange(Number(node.value), type.valueRange)
-				) {
-					ctx.err.report(
-						localize(
-							'number.between',
-							type.valueRange.min ?? '-∞',
-							type.valueRange.max ?? '+∞',
-						),
-						node,
-						core.ErrorSeverity.Warning,
+				if (node.type === 'json:object') {
+					return node.children.filter(kvp => kvp.key).map(kvp => ({
+							key: { originalNode: kvp.key!, inferredType: inferType(kvp.key!) },
+							possibleValues: kvp.value ? [{ originalNode: kvp.value, inferredType: inferType(kvp.value) }] : []
+						})
 					)
 				}
-				break
-			case 'dispatcher':
-				node = node as JsonObjectNode
-				// const id = resolveFieldPath(node.parent?.parent, type.index.path)
-				// if (type.index.registry) {
-				// 	if (ExtendableRootRegistry.is(type.index.registry)) {
-				// 		index(type.index.registry, id ? core.ResourceLocation.lengthen(id) : undefined, options)(node, ctx)
-				// 	} else if (id) {
-				// 		index(type.index.registry, core.ResourceLocation.lengthen(id), options)(node, ctx)
-				// 	}
-				// }
-				break
-			case 'list':
-				node = node as JsonArrayNode
-				type = mcdoc.simplifyListType(type)
-				if (
-					type.lengthRange &&
-					!isInRange(node.children.length, type.lengthRange)
-				) {
-					ctx.err.report(
-						localize(
-							'expected',
-							localize(
-								'json.checker.collection.length-between',
-								localizeTag(node.type),
-								type.lengthRange.min ?? '-∞',
-								type.lengthRange.max ?? '+∞',
-							),
-						),
-						node,
-						core.ErrorSeverity.Warning,
-					)
+				return []
+			},
+			reportError: mcdoc.runtime.checker.getDefaultErrorReporter(ctx, (node, err) => {
```

Comment:
I have a usecase in isAssignable checking if a mcdoctype is assignable to another which hooks this method to just write a boolean if it is called.

Technically the functions of the public interface could also return a boolean, but I thought it's a bit redundant, as it's supposed to just report errors.

I also thought that it might be nice to allow for overriding certain error messages, and an implementation of this could also look like this:

```ts
switch (error.kind) {
    case: 'type_mismatch'
         // custom error handling
         break;
     default: defaultErrorReporter(error);
 }
```

Reply:
For comparing two typedefs, we also might want a custom reporter to say something like "Path a.b.c is a string, but expected an integer", though the current error types do not allow access to this information right now.

Reply:
Okay this makese some sense 👍